Don't use ifconfig -f inet:cidr to extract IP/Mask...

rexpo opened the pull request:
... this is only supported by FreeBSD 11 and newer

Make sure to follow and check these boxes before submitting a PR! Thank you.

*My first pull request - please be gentle!*

- [x] Explain the feature

Nice feature added at commit 6556e18 shows IP address / mask at jail startup, but unfortunately only works on FreeBSD 11 and newer.

Rather than using `ifconfig -f inet:cidr`, just get the netmask in hex format from the standard `ifconfig` output, and convert it to cidr (number of bits).

- [x] Read [CONTRIBUTING.md](https://github.com/iocage/iocage/blob/master/CONTRIBUTING.md)

Commit:
Don't use ifconfig -f inet:cidr to extract IP/Mask, this is only supported by FreeBSD 11 and newer

rexpo commented:
This is one way to fix issue #514 

No worries if you'd rather do it another way - at least (I think) I've finally figured out how do this pull request stuff ;)

skarekrow commented:
Haven't forgot this, just no time to test it, sorry! Will do soon(ish)

skarekrow commented:
Looks good! Just a couple travis fixes, then I can merge.

Sorry it took so long, it won't this time ;)

gronke commented on iocage/lib/ioc_start.py:
```diff
@@ -368,12 +368,14 @@ def __start_jail__(self):
+                    hexmask = out.splitlines()[2].split()[3].decode()
+                    maskcidr = sum([bin(int(hexmask, 16)).count("1")])
+
+                    addr = f"{out.splitlines()[2].split()[1].decode()}/{maskcidr}"
```
For improved readability this could be written as

```
ip4_addr = out.splitlines()[2].split()[1].decode()
addr = f"{ip4_addr}/{maskcidr}"
```

gronke reviewed (dismissed):
Besides the complains of travis this looks good to merge!

gronke commented:
@skarekrow would you fork this branch and rebase it? I'd like to see a travis build pass first, but in general it looks good to merge.

rexpo commented:
Did this ever get merged? I'm still running off of my fork here at home, it's only now that I finally have time to work on iocage again.

Should I make the @gronke readability modification, which I like very much, and do this pull request again (or update it somehow)? Or would it be better if I updated my fork with a fresh HEAD first?

skarekrow commented:
@rexpo The suggest by Stefan is to taste, you can leave it as is if you want. The more critical part is the travis failures and the conflicting files, those can be fixed and rebased in this PR. There isn't a need for a new one unless you find it easier.

rexpo commented:
Well this is a good opportunity for me to try this rebasing stuff.

Holy crap a lot has changed since March! 😄 Let me just replicate my junk over to my spare server before I "convert the configurations to the new format!"... With luck, I'll have this cleaned up and ready for merging in a day or two.

skarekrow commented:
:D

On Fri, Sep 14, 2018, 6:45 PM rexpo <notifications@github.com> wrote:

> Well this is a good opportunity for me to try this rebasing stuff.
>
> Holy crap a lot has changed since March! 😄 Let me just replicate my junk
> over to my spare server before I "convert the configurations to the new
> format!"... With luck, I'll have this cleaned up and ready for merging in a
> day or two.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/iocage/iocage/pull/515#issuecomment-421511760>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABg_N7J1hHpfu-i_55FjErYpu2WaJqPUks5ubD-xgaJpZM4TBtfw>
> .
>
-- 
- Brandon

rexpo commented:
tada! 😃 

Jebus the git is powerful! It took many attempts to get it done, but I certainly learned a lot 😄 And that Travis thingy of yours seems to be happy now. Such a cool system!

rexpo commented:
That is to say, I've rebased and squashed - My fork is now 1 commit ahead of master, with the fix implemented (using the nice @gronke readability suggestion).

Ready for prime time.

☺️

Commit:
efficiencyfied a redundant operation, reordered for readability, and added comments

skarekrow commented on iocage_lib/ioc_start.py:
```diff
@@ -434,12 +434,17 @@ def __start_jail__(self):
                     out = su.check_output(cmd)
 
-                    addr = f"{out.splitlines()[2].split()[1].decode()}"
+                    hexmask = out.splitlines()[2].split()[3].decode()
```
Just noticed the duplication here. Let's make out.splitlines().... A variable called addr_split and have hexmask and ip4_addr use that by index with their decode. Saves us an operation then.

rexpo commented:
Like that, for example? Ready for the debutante ball? 💃

If you bless it, I'll squash again before you merge.

Commit:
Moved the [2] up to original assignment of addr_split

rexpo commented:
Thanks for your patience. It's a bit of a learning curve, but I like it!

I see that someone has committed something to master in the meantime, so a rebase (and squash of course) will be needed before my fix can be committed to master. Is it standard procedure that us slaves do that rebase/squash, so you don't have to? Anyway, I'd like to do that - I need the practice!

skarekrow commented:
It's usually automatic ;)

rexpo commented:
So fancy!

I've spent the weekend migrating all my captive 9.3-RELEASE DHCP VIMAGE jails from my old FreeNAS 9 box to my 10.4-RELEASE iocage lab box. The process goes so smoothly, and the jails seem happy in their new home (thanks to this beautiful patch 😉) And now that they've been liberated from the evil Warden, I can choose to ~use~ test the magical iocage update/upgrade facilities on them ☺️

I've been waiting a year and a half for a nice jail manager that could do this, and now we have it! Well done, team! 👏

Now it's time to celebrate my first ever commit with a tasty drink 🍺

skarekrow commented:
Thanks @rexpo, don't forget it will be in the latest FreeNAS releases, great excuse to upgrade :D
